internal/server/recon/recognize.go

```diff
@@ -73,6 +77,189 @@ func RecognizePlaces(
+	}
+	span := spans[0]
+	queryParts := splitQueryBySpan(query, span)
+	for _, part := range queryParts {
+		if part == span {
+			queryRespItems = append(queryRespItems, span2item[span])
+		} else {
+			queryRespItems = append(queryRespItems, getItemsForSpans(spans[1:], part, span2item)...)
+		}
+	}
+	return queryRespItems
+}
+
+// Splits a query by a span into a list of parts like: non-span part, span part,
+// non-span part, span part, etc. If the query does not contain the span,
+// returns a list with the query as the only item.
+// Examples
+// query: "ab cd ef ab g", span: "ab", result: ["ab", "cd ef", "ab", "g"]
+// query: "ab cd ef g", span: "jk", result: ["ab cd ef g"]
+func splitQueryBySpan(query string, span string) []string {
+	splits := strings.Split(query, span)
+	// if span is not found in query, return just the query
+	if len(splits) < 2 {
+		return []string{query}
+	}
+	queryParts := []string{}
+	// prefix to add to the next split to be processed
+	nextPartPrefix := ""
+	// go through all the splits and add the span between each split if it was a
+	// valid split (we only want to split on complete words)
+	for i, split := range splits {
+		queryPart := nextPartPrefix + split
+		// if this is the last item in the list, add the query part and finish
+		if i == len(splits)-1 {
+			queryParts = append(queryParts, strings.TrimSpace(queryPart))
+			break
+		}
+		// Whether or not to skip adding anything in this round. We want to skip if
+		// this was not a valid split.
+		shouldSkipAdding := false
+		// if the current split is not empty and does not end in a space, assume
+		// span is part of the last word in current split which is not a valid split.
+		if len(split) > 0 && !strings.HasSuffix(split, " ") {
+			shouldSkipAdding = true
+		}
+		// if the next split is not empty and does not start with a space, assume
+		// span is part of the first word in next split which is not a valid split.
+		if len(splits[i+1]) > 0 && !strings.HasPrefix(splits[i+1], " ") {
+			shouldSkipAdding = true
+		}
+		// if we're skipping adding in this round, add span to the current query
+		// part and set that as the prefix for the next part and continue to next
+		// part
+		if shouldSkipAdding {
+			nextPartPrefix = queryPart + span
+			continue
+		}
+		// add current query part and span to the list of query parts.
+		queryParts = append(queryParts, strings.TrimSpace(queryPart))
+		queryParts = append(queryParts, span)
+		nextPartPrefix = ""
+	}
+	return queryParts
+}
+
+// Gets the reconName for a span.
+// The logic here should correspond to the logic for processing name into
+// reconName in flume (https://source.corp.google.com/piper///depot/google3/datacommons/prophet/flume_generator/triple_helper.cc;l=168-173)
+func getReconName(span string) string {
+	reconName := strings.ReplaceAll(span, " ,", "")
+	reconName = strings.ReplaceAll(reconName, ",", "")
+	reconName = strings.ReplaceAll(reconName, "^", "")
+	reconName = strings.ToLower(reconName)
```

Comment:
We could also probably clean up consecutive spaces.  Add TODO?

Reply:
added TODO